### ISSUE TITLE
Clean working directory sometimes fails on MacOS with GoCD 21.3

### ISSUE THREAD

**Issue by sandro-h** [gocd/gocd#9927]
##### Issue Type

- Bug Report

##### Summary

Clean working directory fails in some situations on MacOS.

It looks like this is a regression in commons-io 2.11: https://issues.apache.org/jira/projects/IO/issues/IO-751
GoCD bumped this dependency to 2.11 with the 21.3 release.

##### Environment

- GoCD 21.3
- GoCD agent on MacOS 11.5.2 (Big Sur) 

##### Steps to Reproduce

1. Run a build on a MacOS agent - the exact triggering factors are a hard to determine, most likely when the build creates symlinks in the build output.
2. Run the build again on the same agent

##### Expected Results

The second build cleans up the working directory and builds successfully.

##### Actual Results

The second build fails with:
```
Clean working directory is set to true. Unable to clean working directory for agent: /Users/go/go-agent/pipelines/my-pipeline, with error: pipelines/my-pipeline
```

Looking at the checkout directory, there are a number of folders which are suddenly set to read-only, which a moment before the second build were still writable. So it appears the clean action makes them read-only. This matches the behavior described on the commons-io ticket.

##### Possible Fix

Downgrade commons-io or wait for the fix in https://issues.apache.org/jira/projects/IO/issues/IO-751

**Comment by chadlwilson** [comment]
Thanks for the report and the digging. Do you perhaps have a simple reproducing build task script of basic shell commands, perhaps?

Are you able to get it to reliably fail with a given repository + build process? Does it create symlinks as part of the build tasks?

The Commons IO report is a bit confusing to me, as the user mentions macOS specifically, but then seems to focus on things that would be an issue on all POSIX file systems. I suspect if it were an issue on Linux we'd have a lot of reports, as I daresay that is the OS most folks are running agents with. However perhaps it's something to do with the underlying implementation within APFS or whatever.

**Comment by sandro-h** [comment]
I tested it with older commons-io versions:
* 2.11, 2.10, 2.9.0 all produce the error
* 2.8 works

**Comment by chadlwilson** [comment]
Cool - thanks a lot, I'll try and take a look. It seems some kind of attempt to address this might be planned for `2.12.0` based on [this commit](https://github.com/apache/commons-io/commit/62925e3084c9e72ae4659c00364a490252368326#diff-480762ed7e0cb6175cb14bd289b7167eeac7ae35cf253725155eed73398da621). Have sought some clarity on the original ticket (timing of eventual release is another question mark...).

**Comment by alicekaerast** [comment]
I am not convinced that this is limited to MacOS, as we have very similar-looking behaviour on Amazon Linux 2 with an EFS volume shared between agents.

**Comment by sandro-h** [comment]
I see there's been some activity in the commons-io ticket. Thanks for pursuing that, @chadlwilson. 

I did my little test with the latest commons-io 2.12.0-SNAPSHOT and can confirm that it no longer fails. Apologies for using you as a "proxy" and not answering in the JIRA ticket directly, but I thought perhaps you want to do other or more thorough tests in the full GoCD context, so I didn't want to hijack the conversation with my own limited findings.

As the upstream release looks to be delayed a bit, I will probably downgrade commons-io to 2.8.0 in our GoCD installation, so our MacOS build users aren't blocked. Seems to work without issues based on a test in the staging environment.

**Comment by chadlwilson** [comment]
Nah, that's totally fine - thanks a lot for doing a test!

Any particular reason for going as far as `2.6` rather than `2.8.0`?

**Comment by chadlwilson** [comment]
Part of the reason I am reticent to downgrade is that there is at least one reported fixes for Java 16/17 support in 2.9.0+ versions, such support which I am actively trying to resolve, and I'd rather not work backwards to figure out whether not having those Java 16/17 fixes affects any of the uses within GoCD.

But at this stage given the distractions of the log4shell incident for the Apache folks, it might be a more pragmatic approach. I think we can either downgrade or upgrade commons-io for the next release, depending on whether we get GoCD stable on Java 17 prior to a 2.12 commons-io release.

**Comment by sandro-h** [comment]
I understand the hesitancy to downgrade. Personally I favor moving forward instead of backwards with the next release.

> Any particular reason for going as far as 2.6 rather than 2.8.0?

I based that on your https://github.com/gocd/gocd/issues/9397#issuecomment-880813088 regarding the reason the 2.8 bump was reverted in GoCD ;-) While we don't use Perforce, it sounded like a general problem with deleting read-only files. I see that as a risk of disrupting yet other users which might produce some read-only files in their builds (we have a large variety of projects and builds).

**Comment by chadlwilson** [comment]
Oh, haha - my bad 🤦‍♂️ Bad menory on my part, was thinking 2.8 was the one I initially bumped to. But, yes, that version is problematic. And so is 2.6 (the vuln), and 2.9 to 2.11 ... argh.

**Comment by chadlwilson** [comment]
Thanks to your example I was able to reproduce the error in a unit test, and also validate 2.12.0 appears to fix the specific symlinked-dirs issue. Unfortunately currently there is a regression in at least one other place GoCD uses commons-io which is causing test failures and needs to be addressed before 2.12 is released.

So unless we find another approach/library to clean working dirs that offers the same functionality we might be temporarily stuck (or need to implement yet more workarounds which I have limited time/appetite for) 😓
